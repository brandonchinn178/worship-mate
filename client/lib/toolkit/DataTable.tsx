import _ from 'lodash'
import { Key, ReactNode } from 'react'
import styled, { css } from 'styled-components'

type DataTableProps<T> = {
  // The data to render in the DataTable
  data: T[]
  // The definitions for the columns in the DataTable
  columnDefs: ColumnDefs<T>

  // The key to use for each row (default: the index of the row)
  rowKey?: (rowData: T) => Key
}

export type ColumnDefs<T> = Array<ColumnDef<T>>

type ColumnDef<T> = {
  // The name of the column
  name: string
  // The name of the column to display in the header row (default: `name` capitalized)
  header?: string
  // The size of the column (default: 1fr)
  size?: string
  // A custom rendering function, if desired
  render?: (rowData: T) => ReactNode
}

const renderHeader = <T,>({ name, header }: ColumnDef<T>): ReactNode => {
  if (header) {
    return header
  }

  return _.capitalize(name)
}

const renderCell = <T,>(
  { name, render }: ColumnDef<T>,
  rowData: T,
): ReactNode => {
  if (render) {
    return render(rowData)
  }

  if (typeof rowData === 'object') {
    // TODO: fix?
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (rowData as any)[name]
  }

  return null
}

export default function DataTable<T>({
  data,
  columnDefs,
  rowKey,
}: DataTableProps<T>) {
  // Let columns shrink to dictated proportions, without respect to
  // the data in the columns.
  // https://github.com/rachelandrew/cssgrid-ama/issues/25
  const columnSizes = columnDefs
    .map(({ size }) => size || 'minmax(0, 1fr)')
    .join(' ')

  const getRowKey = (rowData: T, i: number) => (rowKey ? rowKey(rowData) : i)

  return (
    <Table>
      <TableRow columnSizes={columnSizes}>
        {columnDefs.map((columnDef, i) => (
          <TableHeaderCell key={i}>{renderHeader(columnDef)}</TableHeaderCell>
        ))}
      </TableRow>
      {data.map((rowData, i) => (
        <TableRow key={getRowKey(rowData, i)} columnSizes={columnSizes}>
          {columnDefs.map((columnDef, j) => (
            <TableCell key={j}>{renderCell(columnDef, rowData)}</TableCell>
          ))}
        </TableRow>
      ))}
    </Table>
  )
}

const borderStyle = css`1px solid ${(p) => p.theme.colors.black}`

const Table = styled.div`
  display: grid;
  border-top: ${borderStyle};
  border-left: ${borderStyle};
`

const TableRow = styled.div<{ columnSizes: string }>`
  display: grid;
  grid-template-columns: ${(p) => p.columnSizes};
  border-right: ${borderStyle};
  border-bottom: ${borderStyle};
`

const TableCell = styled.div`
  display: grid;
  padding: 5px 10px;
`

const TableHeaderCell = styled(TableCell)`
  ${(p) => p.theme.fonts.label};
  font-size: 1.2rem;
  text-align: center;
`
