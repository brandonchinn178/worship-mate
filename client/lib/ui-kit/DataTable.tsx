import * as _ from 'lodash'
import { Key, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { color, font } from '~/theme'

export type DataTableProps<T> = {
  // The data to render in the DataTable
  data: readonly T[]
  // The definitions for the columns in the DataTable
  columnDefs: ColumnDefs<T>

  // The key to use for each row (default: the index of the row)
  rowKey?: (rowData: T) => Key
}

export type ColumnDefs<T> = Array<ColumnDef<T>>

type ColumnDef<T> = {
  // The name of the column
  name: string
  // The name of the column to display in the header row (default: `name` capitalized)
  header?: string
  // The size of the column (default: 1fr)
  size?: string
  // A custom rendering function, if desired
  render?: (rowData: T) => ReactNode
}

const renderHeader = <T,>({ name, header }: ColumnDef<T>): ReactNode => {
  if (header !== undefined) {
    return header
  }

  return _.capitalize(name)
}

const renderCell = <T,>(
  { name, render }: ColumnDef<T>,
  rowData: T,
): ReactNode => {
  if (render) {
    return render(rowData)
  }

  return _.get(rowData, name)
}

export function DataTable<T>({ data, columnDefs, rowKey }: DataTableProps<T>) {
  // Let columns shrink to dictated proportions, without respect to
  // the data in the columns.
  // https://github.com/rachelandrew/cssgrid-ama/issues/25
  const columnSizes = columnDefs
    .map(({ size }) => size || 'minmax(0, 1fr)')
    .join(' ')

  const getCellKey = (rowData: T, i: number, j: number) => {
    const row = rowKey ? rowKey(rowData) : i
    const col = j
    return `${row}-${col}`
  }

  return (
    <Table columnSizes={columnSizes}>
      <>
        {columnDefs.map((columnDef, i) => (
          <TableHeaderCell key={i}>{renderHeader(columnDef)}</TableHeaderCell>
        ))}
      </>
      {data.map((rowData, i) =>
        columnDefs.map((columnDef, j) => (
          <TableCell key={getCellKey(rowData, i, j)}>
            {renderCell(columnDef, rowData)}
          </TableCell>
        )),
      )}
    </Table>
  )
}

const borderStyle = css`1px solid ${color('black')}`

const Table = styled.div<{ columnSizes: string }>`
  display: grid;
  grid-template-columns: ${(p) => p.columnSizes};

  border: ${borderStyle};
  border-bottom: 0;
  & > * {
    border-bottom: ${borderStyle};
  }
`

const TableCell = styled.div`
  display: grid;
  padding: 10px;
  justify-items: center;
  text-align: center;
  min-width: max-content;
`

const TableHeaderCell = styled(TableCell)`
  padding: 5px;
  ${font('label')}
  font-size: 1.4rem;
  text-align: center;
`
