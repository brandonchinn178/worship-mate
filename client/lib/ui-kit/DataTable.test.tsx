import { renderUI } from '~jest-utils'

import { DataTable } from './DataTable'

type MockPerson = {
  id: number
  name: string
}

const mockData: MockPerson[] = [
  { id: 1, name: 'Alice' },
  { id: 2, name: 'Bob' },
  { id: 3, name: 'Charlie' },
]

it('displays data', () => {
  const columnDefs = [{ name: 'id' }, { name: 'name' }]

  const { container, getByText } = renderUI(
    <DataTable data={mockData} columnDefs={columnDefs} />,
  )

  expect(container).toMatchSnapshot()
  expect(getByText('Id')).toBeVisible()
  expect(getByText('Name')).toBeVisible()
  expect(getByText('1')).toBeVisible()
  expect(getByText('Alice')).toBeVisible()
  expect(getByText('2')).toBeVisible()
  expect(getByText('Bob')).toBeVisible()
  expect(getByText('3')).toBeVisible()
  expect(getByText('Charlie')).toBeVisible()
})

it('displays no data', () => {
  const columnDefs = [{ name: 'id' }, { name: 'name' }]

  const { container, getByText } = renderUI(
    <DataTable data={[]} columnDefs={columnDefs} />,
  )

  expect(container).toMatchSnapshot()
  expect(getByText('Id')).toBeVisible()
  expect(getByText('Name')).toBeVisible()
})

it('displays a different header', () => {
  const columnDefs = [{ name: 'name', header: 'First Name' }]

  const { getByText, queryByText } = renderUI(
    <DataTable data={mockData} columnDefs={columnDefs} />,
  )

  expect(getByText('First Name')).toBeVisible()
  expect(queryByText('Name')).toBeNull()
})

it('displays column with a custom render function', () => {
  const columnDefs = [
    {
      name: 'id+name',
      render: ({ id, name }: MockPerson) => `${id}+${name}`,
    },
  ]

  const { getByText } = renderUI(
    <DataTable data={mockData} columnDefs={columnDefs} />,
  )

  expect(getByText('1+Alice')).toBeVisible()
})
