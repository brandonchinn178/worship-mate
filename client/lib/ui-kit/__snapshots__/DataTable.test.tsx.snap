// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays data 1`] = `
<div>
  <div
    class="DataTable__Table-sc-1xhkibe-0 lhcMRB"
  >
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 DataTable__TableHeaderCell-sc-1xhkibe-2 gRqLRZ"
    >
      Id
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 DataTable__TableHeaderCell-sc-1xhkibe-2 gRqLRZ"
    >
      Name
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      1
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      Alice
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      2
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      Bob
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      3
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 foVggC"
    >
      Charlie
    </div>
  </div>
</div>
`;

exports[`displays no data 1`] = `
<div>
  <div
    class="DataTable__Table-sc-1xhkibe-0 lhcMRB"
  >
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 DataTable__TableHeaderCell-sc-1xhkibe-2 gRqLRZ"
    >
      Id
    </div>
    <div
      class="DataTable__TableCell-sc-1xhkibe-1 DataTable__TableHeaderCell-sc-1xhkibe-2 gRqLRZ"
    >
      Name
    </div>
  </div>
</div>
`;
